Automatic checksum fetching for kernel tarballs

conorsch opened the pull request:
Since the build role contains conditional logic for using the "test" or "stable" grsecurity patches, it must also smartly choose the Linux kernel version, otherwise compilation will fail. Rather than hardcoding all SHA256 checksums and updating them manually each time a new patch is released, we can fetch the list of checksums from kernel.org, verify the GPG signature of the list, and extract only the checksum for the kernel version we're targeting.

We can then pass the checksum to the `get_url` task, so the download for the tarball will be skipped if the file on disk matches the checksum. Without the checksum, Ansible will blindly redownload the tarball on subsequent provisioning runs, which is a waste of time and bandwidth.

Some minor out-of-scope changes are included because they were convenient to make alongside the checksum additions:
- Renames the `target_os` variable to the more descriptive `grsecurity_build_strategy`, defaulting to `vanilla`.
- Replaces `shell` calls with `command` calls where possible (as [recommended by docs](http://docs.ansible.com/ansible/shell_module.html#notes)).
- Leverages default variables for most everything (see `defaults/main.yml`).

Still no support for stable patches, but at this point that's trivially implementable with a separate playbook. Will handle stable patch support separately.

conorsch committed:
Simplifies Linux URL variables by reusing base string.

conorsch committed:
Retrieve SHA256 checksums from kernel.org.

Since the Linux kernel version will change often, updating
the checksums for downloading files will become tedious quickly.
Better to read the checksums from kernel.org directly, extract
the appropriate checksum for the kernel version being used,
then download the kernel version with `sha256sum` passed to
the `get_url` task.

We can also verify the GPG signature of the checksum file itself.
Haven't added that yet.

conorsch committed:
Moves GPG import tasks to separate task list.

In preparation for expanding the verification steps,
let's isolate management of the GPG keys used for
verifying signatures.

conorsch committed:
Cleans up GPG import tasks.

conorsch committed:
Verifies GPG signature for checksum file.

conorsch committed:
Prefer `command` to `shell` calls.

As the Ansible docs state:

    If you want to execute a command securely and predictably, it may be
    better to use the [command] module instead. Best practices
    when writing playbooks will follow the trend of using
    [command] unless [shell] is explicitly required. When running
    ad-hoc commands, use your best judgement. To sanitize any
    variables passed to the shell module, you should use "{{ var |
    quote }}" instead of just "{{ var }}" to make sure they don't
    include evil things like semicolons.

Done. At least, throughout the `build-grsec-kernel` role.

conorsch committed:
Fixes verification tasks.

Removed inappropriate `failed_when` parameters that
could have caused false positives.

conorsch committed:
Fixes minor whitespace problems in kernel file fetching.

conorsch committed:
Import grsecurity vars before any downloads.

Vars files for test/stable patches also determine version
of Linux kernel to download, so the vars must be sourced
before downloading either of those. Added the import
as first task in `tasks/main.yml`.

conorsch committed:
Cleans up verification task list.

More readable task names, favor command over shell.

conorsch committed:
Renames target_os var -> grsecurity_build_strategy.

The target_os variable name was confusing and a bit
misleading. Its possible values were:

  * debian (default)
  * ubuntu

which implied that the kernel packages being built
would only work on those two operating systems. Not so.
While it's still true that special steps are performed
to build for an Ubuntu host, the regular ol' vanilla
kernel should work just fine on a variety of distros.
Therefore the allowed options are now:

  * vanilla (default)
  * ubuntu

The door is left open to supporting additional wacky
distro configs by jumping through hoops conditionally.

conorsch committed:
Changes download directory location for building.

The previous value was `/usr/src`, of which the Linux FHS says
"Don't work here with your own projects". There's a separate
entry for `/usr/src/linux`, but that's no good either:

  You should probably use another directory
  when building your own kernel.

About `/usr/local/share`, the guidance is simply
"Source code for locally installed software",
which sounds downright friendly in the context
of the FHS, so let's use that.

ageis commented on ansible/roles/build-grsec-kernel/tasks/compile.yml:
```diff
@@ -6,30 +6,36 @@
 - name: put kernel config
   template:
-    src: config-{{ target_os }}
+    src: config-{{ grsecurity_build_strategy }}
```
You'll need to also rename `config-debian` to `config-vanilla` in the templates directory for this to work. The included configs are critical for now (since they include the configuration of grsec) but maybe in the future we could find a way to automate collecting a base config from the `/boot` directory of the target machine and then automate the running of `make menuconfig` to enable and configure grsec and generate a `.config` file output with the options we want.

conorsch replied:
@ageis Good catch! Rebased to fix, change now incorporated in 17789db.

conorsch commented:
Will open a separate issue for reworking the `compile.yml` flow, particularly so we don't have to depend on `config` files shipped with the role.

conorsch committed:
Renames target_os -> grsecurity_build_strategy in all roles.

Previously fixed in just the kernel package building role,
but the `target_os` var was used in the other roles, too.
Updated the defaults for each role, and retained the top-level
vars file for exposing a single point for updating the value.

conorsch committed:
Updates grsecurity URLs with newest releases.

conorsch commented:
Opened #15 to focus on changes to the compile process.

ageis commented:
In the build role, `roles/build-grsec-kernel/tasks/copy.yml` should be updated so the package is fetched from 

```
src={{ download_directory }}/{{ package_filename }}
```

it still has `/usr/src` hardcoded. As soon as that's resolved, this stuff has been tested by me and is good to merge.

conorsch commented:
Tracking changes to compile-and-fetch process in #15, thanks for the careful review!
